File: TEAM B - Afternoon/Week 3/Madeline/2805.py
# 2805 나무 자르기
# 나무 M미터
# 목재 절단기 - 높이 H 지정해서 자름(양의 정수, 0)
# ex) 한 줄에 있는 나무가 20, 15, 10, 17이고, 높이가 15라면,
# 잘린 나무의 높이: 15, 15, 10, 15
# -> 상근이는 5,2 나무를 들고 집에 간다.
# => 최소한 M 미터를 들고 집에 갈거야. 높이의 최댓값을 구해

import sys

input = sys.stdin.readline

#이진탐색
# 높이 = target
def BSearch(tree, first, last, M):
    while first <= last:
        target = 0
        mid = int((first + last) / 2)
        for i in tree:
            if(i > mid):
                target += i - mid
        if target >= M:
            first = mid + 1
        else:
            last = mid - 1
    return last


# N개의 나무, M의 나무를 집에 가져가
N, M = map(int, input().split())

tree = list(map(int, input().split()))

print(BSearch(tree, 1, max(tree), M))

File: TEAM B - Morning (Season 1)/Week 5-이진탐색/Ddan/[백준 2805]나무 자르기/2805.swift
//
//  2805.swift
//  AlgorithmStudy
//
//  Created by 박승찬 on 2023/06/23.
//

import Foundation

//func solution2805() {
//    let NM: [Int] = readLine()!.components(separatedBy: " ").compactMap{Int(String($0))!}
//    let trees: [Int] = readLine()!.components(separatedBy: " ").compactMap{Int(String($0))!}
//
//    var start: Int = 0
//    var end: Int = trees.max()!
//    var middle = 0
//    var result = 0
//    while start <= end {
//        middle = (start + end) / 2
//        result = 0
//        result = trees.map { $0 > middle ? $0 - middle : 0 }.reduce(0, +)
//        if result >= NM[1] {
//            start = middle + 1
//        } else {
//            end = middle - 1
//        }
//    }
//    print(start-1)
//}


func solution2805() {
    let input = readLine()!.split(separator: " ").compactMap{ Int(String($0))}
    let m = input[1]
    let trees = readLine()!.split(separator: " ").compactMap{ Int(String($0)) }

    var min = 0
    var max = trees.max()!
    var mid = 0
    var result = 0
    var temp = 0

    while min < max {
      mid = (min + max) / 2
      temp = trees.map { $0 > mid ? $0 - mid : 0 }.reduce(0, +)
      
      if(temp < m) {
        max = mid
      } else {
        result = mid
        min = mid + 1
      }
    }

    print(result)
}

